File: platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/annotations/DraggableAnnotationControllerTest.java
package org.maplibre.android.annotations;

import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertTrue;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import android.graphics.PointF;

import com.mapbox.android.gestures.AndroidGesturesManager;
import com.mapbox.android.gestures.MoveDistancesObject;
import com.mapbox.android.gestures.MoveGestureDetector;
import com.mapbox.geojson.Geometry;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.maplibre.android.maps.MapLibreMap;
import org.maplibre.android.maps.MapView;
import org.maplibre.android.maps.Projection;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import java.util.ArrayList;
import java.util.List;

@RunWith(RobolectricTestRunner.class)
public class DraggableAnnotationControllerTest {

  private static final int touchAreaMaxX = 100;
  private static final int touchAreaMaxY = 100;

  @Mock
  private MapView mapView;

  @Mock
  private MapLibreMap maplibreMap;

  @Mock
  private Projection projection;

  @Mock
  private AndroidGesturesManager androidGesturesManager;

  @Mock
  private MoveGestureDetector moveGestureDetector;

  @Mock
  private MoveDistancesObject moveObject;

  @Mock
  private AnnotationManager annotationManager;

  @Mock
  private AbstractAnnotation annotation;

  @Mock
  private Geometry geometry;

  @Mock
  private OnAnnotationDragListener dragListener;

  private List<OnAnnotationDragListener> dragListenerList;

  private DraggableAnnotationController draggableAnnotationController;

  @Before
  public void before() {
    MockitoAnnotations.initMocks(this);
    draggableAnnotationController = new DraggableAnnotationController(mapView, maplibreMap, androidGesturesManager,
      0, 0, touchAreaMaxX, touchAreaMaxY);
    draggableAnnotationController.addAnnotationManager(annotationManager);
    dragListenerList = new ArrayList<>();
    dragListenerList.add(dragListener);
  }

  @Test
  public void annotationNotDraggableTest() {
    when(annotation.isDraggable()).thenReturn(false);
    draggableAnnotationController.startDragging(annotation, annotationManager);
    verify(dragListener, times(0)).onAnnotationDragStarted(annotation);
  }

  @Test
  public void annotationDragStartOnAdditionalManagerTest() {
    AnnotationManager additionalAnnotationManager = mock(AnnotationManager.class);
    draggableAnnotationController.addAnnotationManager(additionalAnnotationManager);

    when(annotation.isDraggable()).thenReturn(true);
    when(additionalAnnotationManager.getDragListeners()).thenReturn(dragListenerList);
    draggableAnnotationController.startDragging(annotation, additionalAnnotationManager);
    verify(dragListener, times(1)).onAnnotationDragStarted(annotation);
    draggableAnnotationController.stopDragging(annotation, additionalAnnotationManager);
    verify(dragListener, times(1)).onAnnotationDragFinished(annotation);

    draggableAnnotationController.removeAnnotationManager(additionalAnnotationManager);
  }

  @Test
  public void annotationDragStartTest() {
    when(annotation.isDraggable()).thenReturn(true);
    when(annotationManager.getDragListeners()).thenReturn(dragListenerList);
    draggableAnnotationController.startDragging(annotation, annotationManager);
    verify(dragListener, times(1)).onAnnotationDragStarted(annotation);
  }

  @Test
  public void annotationDragStopNoneTest() {
    draggableAnnotationController.stopDragging(null, null);
    verify(dragListener, times(0)).onAnnotationDragFinished(annotation);
  }

  @Test
  public void annotationDragStopTest() {
    when(annotation.isDraggable()).thenReturn(true);
    when(annotationManager.getDragListeners()).thenReturn(dragListenerList);
    draggableAnnotationController.stopDragging(annotation, annotationManager);
    verify(dragListener, times(1)).onAnnotationDragFinished(annotation);
  }

  @Test
  public void annotationDragStopOnDeleteTest() {
    when(annotation.isDraggable()).thenReturn(true);
    when(annotationManager.getDragListeners()).thenReturn(dragListenerList);
    draggableAnnotationController.startDragging(annotation, annotationManager);
    draggableAnnotationController.onAnnotationDeleted(annotation);
    verify(dragListener, times(1)).onAnnotationDragFinished(annotation);
  }

  @Test
  public void gestureOnMoveBeginWrongPointersTest() {
    when(annotation.isDraggable()).thenReturn(true);
    when(annotationManager.getDragListeners()).thenReturn(dragListenerList);

    PointF pointF = new PointF();
    when(annotationManager.queryMapForFeatures(pointF)).thenReturn(annotation);
    when(moveGestureDetector.getFocalPoint()).thenReturn(pointF);
    when(moveGestureDetector.getPointersCount()).thenReturn(0);

    boolean moveBegan1 = draggableAnnotationController.onMoveBegin(moveGestureDetector);
    assertFalse(moveBegan1);

    when(moveGestureDetector.getPointersCount()).thenReturn(2);
    boolean moveBegan2 = draggableAnnotationController.onMoveBegin(moveGestureDetector);
    assertFalse(moveBegan2);
    verify(dragListener, times(0)).onAnnotationDragStarted(annotation);
  }

  @Test
  public void gestureOnMoveBeginTest() {
    when(annotation.isDraggable()).thenReturn(true);
    when(annotationManager.getDragListeners()).thenReturn(dragListenerList);

    PointF pointF = new PointF();
    when(annotationManager.queryMapForFeatures(pointF)).thenReturn(annotation);
    when(moveGestureDetector.getFocalPoint()).thenReturn(pointF);
    when(moveGestureDetector.getPointersCount()).thenReturn(1);

    boolean moveBegan = draggableAnnotationController.onMoveBegin(moveGestureDetector);
    assertTrue(moveBegan);
    verify(dragListener, times(1)).onAnnotationDragStarted(annotation);
  }

  @Test
  public void gestureOnMoveMoveWrongPointersTest() {
    when(annotation.isDraggable()).thenReturn(true);
    when(annotationManager.getDragListeners()).thenReturn(dragListenerList);
    draggableAnnotationController.startDragging(annotation, annotationManager);

    when(moveGestureDetector.getPointersCount()).thenReturn(2);

    boolean moved = draggableAnnotationController.onMove(moveGestureDetector);
    assertTrue(moved);
    verify(dragListener, times(0)).onAnnotationDrag(annotation);
    verify(dragListener, times(1)).onAnnotationDragFinished(annotation);
  }

  @Test
  public void gestureOnMoveBeginNonDraggableAnnotationTest() {
    when(annotation.isDraggable()).thenReturn(false);
    when(annotationManager.getDragListeners()).thenReturn(dragListenerList);

    PointF pointF = new PointF();
    when(annotationManager.queryMapForFeatures(pointF)).thenReturn(annotation);
    when(moveGestureDetector.getFocalPoint()).thenReturn(pointF);
    when(moveGestureDetector.getPointersCount()).thenReturn(1);

    boolean moveBegan = draggableAnnotationController.onMoveBegin(moveGestureDetector);
    assertFalse(moveBegan);
    verify(dragListener, times(0)).onAnnotationDragStarted(annotation);
  }

  @Test
  public void gestureOnMoveMoveOutOfBoundsTest() {
    when(annotation.isDraggable()).thenReturn(true);
    when(annotationManager.getDragListeners()).thenReturn(dragListenerList);
    draggableAnnotationController.startDragging(annotation, annotationManager);

    when(moveGestureDetector.getPointersCount()).thenReturn(1);
    when(moveGestureDetector.getMoveObject(0)).thenReturn(moveObject);

    when(moveObject.getCurrentX()).thenReturn(touchAreaMaxX + 1f);
    when(moveObject.getCurrentY()).thenReturn(10f);
    draggableAnnotationController.onMove(moveGestureDetector);

    when(moveObject.getCurrentX()).thenReturn(10f);
    when(moveObject.getCurrentY()).thenReturn(touchAreaMaxY + 1f);
    draggableAnnotationController.startDragging(annotation, annotationManager);
    draggableAnnotationController.onMove(moveGestureDetector);

    when(moveObject.getCurrentX()).thenReturn(-1f);
    when(moveObject.getCurrentY()).thenReturn(10f);
    draggableAnnotationController.startDragging(annotation, annotationManager);
    draggableAnnotationController.onMove(moveGestureDetector);

    when(moveObject.getCurrentX()).thenReturn(10f);
    when(moveObject.getCurrentY()).thenReturn(-1f);
    draggableAnnotationController.startDragging(annotation, annotationManager);
    draggableAnnotationController.onMove(moveGestureDetector);

    verify(dragListener, times(0)).onAnnotationDrag(annotation);
    verify(dragListener, times(4)).onAnnotationDragFinished(annotation);
  }

  @Test
  public void gestureOnMoveMoveNoGeometryTest() {
    when(annotation.isDraggable()).thenReturn(true);
    when(annotationManager.getDragListeners()).thenReturn(dragListenerList);
    draggableAnnotationController.startDragging(annotation, annotationManager);

    when(moveGestureDetector.getPointersCount()).thenReturn(1);
    when(moveGestureDetector.getMoveObject(0)).thenReturn(moveObject);

    when(moveObject.getCurrentX()).thenReturn(10f);
    when(moveObject.getCurrentY()).thenReturn(10f);

    when(maplibreMap.getProjection()).thenReturn(projection);
    when(annotation.getOffsetGeometry(projection, moveObject, 0, 0)).thenReturn(null);

    boolean moved = draggableAnnotationController.onMove(moveGestureDetector);

    assertFalse(moved);
    verify(dragListener, times(0)).onAnnotationDrag(annotation);
  }

  @Test
  public void gestureOnMoveTest() {
    when(annotation.isDraggable()).thenReturn(true);
    when(annotationManager.getDragListeners()).thenReturn(dragListenerList);
    draggableAnnotationController.startDragging(annotation, annotationManager);

    when(moveGestureDetector.getPointersCount()).thenReturn(1);
    when(moveGestureDetector.getMoveObject(0)).thenReturn(moveObject);

    when(moveObject.getCurrentX()).thenReturn(10f);
    when(moveObject.getCurrentY()).thenReturn(10f);

    when(maplibreMap.getProjection()).thenReturn(projection);
    when(annotation.getOffsetGeometry(projection, moveObject, 0, 0)).thenReturn(geometry);

    boolean moved = draggableAnnotationController.onMove(moveGestureDetector);

    assertTrue(moved);
    verify(annotation).setGeometry(geometry);
    verify(annotationManager).updateSource();
    verify(dragListener, times(1)).onAnnotationDrag(annotation);
  }

  @Test
  public void gestureOnMoveEndTest() {
    when(annotation.isDraggable()).thenReturn(true);
    when(annotationManager.getDragListeners()).thenReturn(dragListenerList);
    draggableAnnotationController.startDragging(annotation, annotationManager);

    draggableAnnotationController.onMoveEnd();
    verify(dragListener, times(1)).onAnnotationDragFinished(annotation);
  }

  @Test
  public void startedNotDraggableTest() {
    when(annotation.isDraggable()).thenReturn(false);
    when(annotationManager.getDragListeners()).thenReturn(dragListenerList);
    draggableAnnotationController.startDragging(annotation, annotationManager);

    when(moveGestureDetector.getPointersCount()).thenReturn(1);
    when(moveGestureDetector.getMoveObject(0)).thenReturn(moveObject);

    when(moveObject.getCurrentX()).thenReturn(10f);
    when(moveObject.getCurrentY()).thenReturn(10f);

    when(maplibreMap.getProjection()).thenReturn(projection);
    when(annotation.getOffsetGeometry(projection, moveObject, 0, 0)).thenReturn(geometry);

    boolean moved = draggableAnnotationController.onMove(moveGestureDetector);

    assertFalse(moved);
    verify(dragListener, times(0)).onAnnotationDrag(annotation);
  }

  @Test
  public void moveNotDraggableTest() {
    when(annotation.isDraggable()).thenReturn(true);
    when(annotationManager.getDragListeners()).thenReturn(dragListenerList);
    draggableAnnotationController.startDragging(annotation, annotationManager);

    when(moveGestureDetector.getPointersCount()).thenReturn(1);
    when(moveGestureDetector.getMoveObject(0)).thenReturn(moveObject);

    when(moveObject.getCurrentX()).thenReturn(10f);
    when(moveObject.getCurrentY()).thenReturn(10f);

    when(maplibreMap.getProjection()).thenReturn(projection);
    when(annotation.getOffsetGeometry(projection, moveObject, 0, 0)).thenReturn(geometry);

    when(annotation.isDraggable()).thenReturn(false);
    boolean moved = draggableAnnotationController.onMove(moveGestureDetector);

    assertTrue(moved);
    verify(dragListener, times(0)).onAnnotationDrag(annotation);
  }
}

File: platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/annotations/DraggableAnnotationControllerTest.kt
package org.maplibre.android.annotations

import android.graphics.PointF
import com.mapbox.android.gestures.AndroidGesturesManager
import com.mapbox.android.gestures.MoveDistancesObject
import com.mapbox.android.gestures.MoveGestureDetector
import com.mapbox.geojson.Geometry
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.maplibre.android.maps.MapLibreMap
import org.maplibre.android.maps.MapView
import org.maplibre.android.maps.Projection
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.MockitoAnnotations
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class DraggableAnnotationControllerTest {
    @Mock
    private lateinit var mapView: MapView

    @Mock
    private lateinit var maplibreMap: MapLibreMap

    @Mock
    private lateinit var projection: Projection

    @Mock
    private lateinit var androidGesturesManager: AndroidGesturesManager

    @Mock
    private lateinit var moveGestureDetector: MoveGestureDetector

    @Mock
    private lateinit var moveObject: MoveDistancesObject

    @Mock
    private lateinit var annotationManager: AnnotationManager<*, Symbol>

    @Mock
    private lateinit var annotation: Symbol

    private lateinit var dragPair: DraggableAnnotationController.DragPair<*>

    @Mock
    private lateinit var dragListener: OnAnnotationDragListener<Symbol>
    private lateinit var dragListenerList: List<OnAnnotationDragListener<Symbol>>
    private lateinit var draggableAnnotationController: DraggableAnnotationController
    @Before
    fun before() {
        MockitoAnnotations.initMocks(this)
        dragPair = DraggableAnnotationController.DragPair(annotation, annotationManager)
        draggableAnnotationController = DraggableAnnotationController(
            mapView, maplibreMap, androidGesturesManager,
            0, 0, touchAreaMaxX, touchAreaMaxY
        )
        draggableAnnotationController.addAnnotationManager(annotationManager)
        dragListenerList = listOf(dragListener)
    }

    @Test
    fun annotationNotDraggableTest() {
        Mockito.`when`(annotation.draggable).thenReturn(false)
        draggableAnnotationController.startDragging(dragPair)
        Mockito.verify(dragListener, Mockito.times(0)).onAnnotationDragStarted(annotation)
    }

    @Test
    fun annotationDragStartOnAdditionalManagerTest() {
        val additionalAnnotationManager = Mockito.mock(
            AnnotationManager::class.java
        )
        draggableAnnotationController.addAnnotationManager(additionalAnnotationManager)
        Mockito.`when`(annotation.draggable).thenReturn(true)
        // Additional annotation manager is ignored (no way of knowing whether it handles same annotation type)
        Mockito.`when`(additionalAnnotationManager.getDragListeners()).thenReturn(dragListenerList)

        Mockito.`when`(annotationManager.getDragListeners()).thenReturn(dragListenerList)
        draggableAnnotationController.startDragging(dragPair)
        Mockito.verify(dragListener, Mockito.times(1)).onAnnotationDragStarted(annotation)
        draggableAnnotationController.stopDragging(dragPair)
        Mockito.verify(dragListener, Mockito.times(1)).onAnnotationDragFinished(annotation)
        draggableAnnotationController.removeAnnotationManager(additionalAnnotationManager)
    }

    @Test
    fun annotationDragStartTest() {
        Mockito.`when`(annotation.draggable).thenReturn(true)
        Mockito.`when`(annotationManager.getDragListeners()).thenReturn(dragListenerList)
        draggableAnnotationController.startDragging(dragPair)
        Mockito.verify(dragListener, Mockito.times(1)).onAnnotationDragStarted(annotation)
    }

    @Test
    fun annotationDragStopNoneTest() {
        draggableAnnotationController.stopDragging<Symbol>(null)
        Mockito.verify(dragListener, Mockito.times(0)).onAnnotationDragFinished(annotation)
    }

    @Test
    fun annotationDragStopTest() {
        Mockito.`when`(annotation.draggable).thenReturn(true)
        Mockito.`when`(annotationManager.getDragListeners()).thenReturn(dragListenerList)
        draggableAnnotationController.stopDragging(dragPair)
        Mockito.verify(dragListener, Mockito.times(1)).onAnnotationDragFinished(annotation)
    }

    @Test
    fun annotationDragStopOnDeleteTest() {
        Mockito.`when`(annotation.draggable).thenReturn(true)
        Mockito.`when`(annotationManager.getDragListeners()).thenReturn(dragListenerList)
        draggableAnnotationController.startDragging(dragPair)
        draggableAnnotationController.onAnnotationDeleted(annotation)
        Mockito.verify(dragListener, Mockito.times(1)).onAnnotationDragFinished(annotation)
    }

    @Test
    fun gestureOnMoveBeginWrongPointersTest() {
        Mockito.`when`(annotation.draggable).thenReturn(true)
        Mockito.`when`(annotationManager.getDragListeners()).thenReturn(dragListenerList)
        val pointF = PointF()
        Mockito.`when`(annotationManager.queryMapForFeatures(pointF)).thenReturn(annotation)
        Mockito.`when`(moveGestureDetector.focalPoint).thenReturn(pointF)
        Mockito.`when`(moveGestureDetector.pointersCount).thenReturn(0)
        val moveBegan1 = draggableAnnotationController.onMoveBegin(moveGestureDetector)
        Assert.assertFalse(moveBegan1)
        Mockito.`when`(moveGestureDetector.pointersCount).thenReturn(2)
        val moveBegan2 = draggableAnnotationController.onMoveBegin(moveGestureDetector)
        Assert.assertFalse(moveBegan2)
        Mockito.verify(dragListener, Mockito.times(0)).onAnnotationDragStarted(annotation)
    }

    @Test
    fun gestureOnMoveBeginTest() {
        Mockito.`when`(annotation.draggable).thenReturn(true)
        Mockito.`when`(annotationManager.getDragListeners()).thenReturn(dragListenerList)
        val pointF = PointF()
        Mockito.`when`(annotationManager.queryMapForFeatures(pointF)).thenReturn(annotation)
        Mockito.`when`(moveGestureDetector.focalPoint).thenReturn(pointF)
        Mockito.`when`(moveGestureDetector.pointersCount).thenReturn(1)
        val moveBegan = draggableAnnotationController.onMoveBegin(moveGestureDetector)
        Assert.assertTrue(moveBegan)
        Mockito.verify(dragListener, Mockito.times(1)).onAnnotationDragStarted(annotation)
    }

    @Test
    fun gestureOnMoveMoveWrongPointersTest() {
        Mockito.`when`(annotation.draggable).thenReturn(true)
        Mockito.`when`(annotationManager.getDragListeners()).thenReturn(dragListenerList)
        draggableAnnotationController.startDragging(dragPair)
        Mockito.`when`(moveGestureDetector.pointersCount).thenReturn(2)
        val moved = draggableAnnotationController.onMove(moveGestureDetector, 0f, 0f)
        Assert.assertTrue(moved)
        Mockito.verify(dragListener, Mockito.times(0)).onAnnotationDrag(annotation)
        Mockito.verify(dragListener, Mockito.times(1)).onAnnotationDragFinished(annotation)
    }

    @Test
    fun gestureOnMoveBeginNonDraggableAnnotationTest() {
        Mockito.`when`(annotation.draggable).thenReturn(false)
        Mockito.`when`(annotationManager.getDragListeners()).thenReturn(dragListenerList)
        val pointF = PointF()
        Mockito.`when`(annotationManager.queryMapForFeatures(pointF)).thenReturn(annotation)
        Mockito.`when`(moveGestureDetector.focalPoint).thenReturn(pointF)
        Mockito.`when`(moveGestureDetector.pointersCount).thenReturn(1)
        val moveBegan = draggableAnnotationController.onMoveBegin(moveGestureDetector)
        Assert.assertFalse(moveBegan)
        Mockito.verify(dragListener, Mockito.times(0)).onAnnotationDragStarted(annotation)
    }

    @Test
    fun gestureOnMoveMoveOutOfBoundsTest() {
        Mockito.`when`(annotation.draggable).thenReturn(true)
        Mockito.`when`(annotationManager.getDragListeners()).thenReturn(dragListenerList)
        draggableAnnotationController.startDragging(dragPair)
        Mockito.`when`(moveGestureDetector.pointersCount).thenReturn(1)
        Mockito.`when`(moveGestureDetector.getMoveObject(0)).thenReturn(moveObject)
        Mockito.`when`(moveObject.currentX).thenReturn(touchAreaMaxX + 1f)
        Mockito.`when`(moveObject.currentY).thenReturn(10f)
        draggableAnnotationController.onMove(moveGestureDetector, 0f, 0f)
        Mockito.`when`(moveObject.currentX).thenReturn(10f)
        Mockito.`when`(moveObject.currentY).thenReturn(touchAreaMaxY + 1f)
        draggableAnnotationController.startDragging(dragPair)
        draggableAnnotationController.onMove(moveGestureDetector, 0f, 0f)
        Mockito.`when`(moveObject.currentX).thenReturn(-1f)
        Mockito.`when`(moveObject.currentY).thenReturn(10f)
        draggableAnnotationController.startDragging(dragPair)
        draggableAnnotationController.onMove(moveGestureDetector, 0f, 0f)
        Mockito.`when`(moveObject.currentX).thenReturn(10f)
        Mockito.`when`(moveObject.currentY).thenReturn(-1f)
        draggableAnnotationController.startDragging(dragPair)
        draggableAnnotationController.onMove(moveGestureDetector, 0f, 0f)
        Mockito.verify(dragListener, Mockito.times(0)).onAnnotationDrag(annotation)
        Mockito.verify(dragListener, Mockito.times(4)).onAnnotationDragFinished(annotation)
    }

    @Test
    fun gestureOnMoveMoveNoGeometryTest() {
        Mockito.`when`(annotation.draggable).thenReturn(true)
        Mockito.`when`(annotationManager.getDragListeners()).thenReturn(dragListenerList)
        draggableAnnotationController.startDragging(dragPair)
        Mockito.`when`(moveGestureDetector.pointersCount).thenReturn(1)
        Mockito.`when`(moveGestureDetector.getMoveObject(0)).thenReturn(moveObject)
        Mockito.`when`(moveObject.currentX).thenReturn(10f)
        Mockito.`when`(moveObject.currentY).thenReturn(10f)
        Mockito.`when`(maplibreMap.projection).thenReturn(projection)
        Mockito.`when`(annotation.offsetGeometry(projection, moveObject, 0f, 0f)).thenReturn(false)
        val moved = draggableAnnotationController.onMove(moveGestureDetector, 0f, 0f)
        Assert.assertFalse(moved)
        Mockito.verify(dragListener, Mockito.times(0)).onAnnotationDrag(annotation)
    }

    @Test
    fun gestureOnMoveTest() {
        Mockito.`when`(annotation.draggable).thenReturn(true)
        Mockito.`when`(annotationManager.getDragListeners()).thenReturn(dragListenerList)
        draggableAnnotationController.startDragging(dragPair)
        Mockito.`when`(moveGestureDetector.pointersCount).thenReturn(1)
        Mockito.`when`(moveGestureDetector.getMoveObject(0)).thenReturn(moveObject)
        Mockito.`when`(moveObject.currentX).thenReturn(10f)
        Mockito.`when`(moveObject.currentY).thenReturn(10f)
        Mockito.`when`(maplibreMap.projection).thenReturn(projection)
        Mockito.`when`(annotation.offsetGeometry(projection, moveObject, 0f, 0f)).thenReturn(true)
        val moved = draggableAnnotationController.onMove(moveGestureDetector, 0f, 0f)
        Assert.assertTrue(moved)
        Mockito.verify(dragListener, Mockito.times(1)).onAnnotationDrag(annotation)
    }

    @Test
    fun gestureOnMoveEndTest() {
        Mockito.`when`(annotation.draggable).thenReturn(true)
        Mockito.`when`(annotationManager.getDragListeners()).thenReturn(dragListenerList)
        draggableAnnotationController.startDragging(dragPair)
        draggableAnnotationController.onMoveEnd(moveGestureDetector, 0f, 0f)
        Mockito.verify(dragListener, Mockito.times(1)).onAnnotationDragFinished(annotation)
    }

    @Test
    fun startedNotDraggableTest() {
        Mockito.`when`(annotation.draggable).thenReturn(false)
        Mockito.`when`(annotationManager.getDragListeners()).thenReturn(dragListenerList)
        draggableAnnotationController.startDragging(dragPair)
        Mockito.`when`(moveGestureDetector.pointersCount).thenReturn(1)
        Mockito.`when`(moveGestureDetector.getMoveObject(0)).thenReturn(moveObject)
        Mockito.`when`(moveObject.currentX).thenReturn(10f)
        Mockito.`when`(moveObject.currentY).thenReturn(10f)
        Mockito.`when`(maplibreMap.projection).thenReturn(projection)
        Mockito.`when`(annotation.offsetGeometry(projection, moveObject, 0f, 0f)).thenReturn(true)
        val moved = draggableAnnotationController.onMove(moveGestureDetector, 0f, 0f)
        Assert.assertFalse(moved)
        Mockito.verify(dragListener, Mockito.times(0)).onAnnotationDrag(annotation)
    }

    @Test
    fun moveNotDraggableTest() {
        Mockito.`when`(annotation.draggable).thenReturn(true)
        Mockito.`when`(annotationManager.getDragListeners()).thenReturn(dragListenerList)
        draggableAnnotationController.startDragging(dragPair)
        Mockito.`when`(moveGestureDetector.pointersCount).thenReturn(1)
        Mockito.`when`(moveGestureDetector.getMoveObject(0)).thenReturn(moveObject)
        Mockito.`when`(moveObject.currentX).thenReturn(10f)
        Mockito.`when`(moveObject.currentY).thenReturn(10f)
        Mockito.`when`(maplibreMap.projection).thenReturn(projection)
        Mockito.`when`(annotation.offsetGeometry(projection, moveObject, 0f, 0f)).thenReturn(true)
        Mockito.`when`(annotation.draggable).thenReturn(false)
        val moved = draggableAnnotationController.onMove(moveGestureDetector, 0f, 0f)
        Assert.assertTrue(moved)
        Mockito.verify(dragListener, Mockito.times(0)).onAnnotationDrag(annotation)
    }

    companion object {
        private const val touchAreaMaxX = 100
        private const val touchAreaMaxY = 100
    }
}